src/zarr.cpp

```diff
@@ -127,12 +127,11 @@ DeviceState
 zarr_start(Storage* self_) noexcept
 {
     DeviceState state{ DeviceState_AwaitingConfiguration };
-
     try {
         CHECK(self_);
         auto* self = (zarr::Zarr*)self_;
         self->start();
-        state = DeviceState_Running;
+        state = self->state;
```

Comment:
There's not a great reason. `DeviceState_AwaitingConfiguration` is what was previously returned on failure, and not setting the state on, e.g., start and stop was the bug. Nowhere does the Zarr object ever set its own state to `DeviceState_AwaitingConfiguration`, except in the constructor. The contract around state changes could be better documented and enforced. That's how this managed to slip under the radar for so long.

OTOH we only check that the state is `DeviceState_Running` on `storage_start`, and it could be either `DeviceState_Armed` or `DeviceState_AwaitingConfiguration` on `storage_stop`, which means it should be fine to return current device state.

Reply:
Wait, there's a simpler reason. The `self_` pointer could be null. My point about the contract being sporadically enforced still stands though.

Reply:
My understanding of the intended state changes is not great, so this explanation makes sense in the context of this PR, so this is resolved (I.e. we should try to improve and document it in future work).

Reply:
100%